File: Timetable/backend/app/substitution.py
"""Substitute-teacher suggestions for a specific date/teacher absence.

Two ways to get the "lessons" this runs against:
  - the live generated timetable (crud.get_all_lessons), for the normal case
  - a one-off uploaded timetable export (app/timetable_workbook.py's
    parse_generated_workbook()["lessons"]), for a coordinator who wants to
    compute suggestions against a timetable that hasn't been saved to the app
    yet - this is never written to the database, it only exists for the one
    /substitution/suggest request that used it.
"""

import re

# A candidate is bumped to the top tier if they teach DMT/Sports (in any of
# its usual spellings) to the exact same grade+section, even if that's not
# the period they're normally free.
DMT_SPORTS_KEYWORDS = ("dmt", "dance", "music", "theatre", "sport")


def _norm(s):
    return re.sub(r"\s+", " ", (s or "").strip()).lower()


def compute_suggestions(lessons, absent_teacher_name, day_of_week, all_teacher_names):
    """lessons: the whole week's placed lessons (from either source above).
    Returns a list of {period_number, grade_name, section_name, subject,
    suggestions: [{teacher_name, tier, tier_label, periods_today, periods_week}]}
    for every period the absent teacher has on day_of_week, ranked:
      1. DMT/Sports teacher for this exact grade+section, and free then
      2. Teacher of this exact subject for this exact grade+section, and free
      3. Any teacher of this grade (any section), and free
      4. Any other free teacher
    Within a tier, the least-loaded candidate (fewest periods that day, then
    fewest that week) is listed first - periods_today/periods_week are each
    candidate's EXISTING load (not counting the vacant period itself), so a
    coordinator can avoid stacking a substitution onto someone already full.
    """
    absent_key = _norm(absent_teacher_name)

    display_name = {_norm(n): n for n in all_teacher_names if n}
    for l in lessons:
        if l["teacher_name"]:
            display_name.setdefault(_norm(l["teacher_name"]), l["teacher_name"])

    occupied_period = {}       # teacher_key -> set(period_number) busy on day_of_week
    teacher_subjects = {}      # teacher_key -> set(subject, lowercased)
    teacher_grade_sections = {}  # teacher_key -> set((grade_name, section_name))
    teacher_grades = {}        # teacher_key -> set(grade_name)
    teacher_day_count = {}     # teacher_key -> periods already taught on day_of_week
    teacher_week_count = {}    # teacher_key -> periods already taught across the whole week

    for l in lessons:
        if not l["teacher_name"]:
            continue
        key = _norm(l["teacher_name"])
        teacher_subjects.setdefault(key, set()).add(_norm(l["subject"]))
        teacher_grade_sections.setdefault(key, set()).add((l["grade_name"], l["section_name"]))
        teacher_grades.setdefault(key, set()).add(l["grade_name"])
        teacher_week_count[key] = teacher_week_count.get(key, 0) + 1
        if l["day_of_week"] == day_of_week:
            occupied_period.setdefault(key, set()).add(l["period_number"])
            teacher_day_count[key] = teacher_day_count.get(key, 0) + 1

    absent_periods = sorted(
        (l for l in lessons if _norm(l["teacher_name"]) == absent_key and l["day_of_week"] == day_of_week),
        key=lambda l: l["period_number"],
    )

    all_keys = set(display_name.keys()) - {absent_key}

    def build(tier_keys, tier_num, label):
        return [
            {
                "teacher_name": display_name[k], "tier": tier_num, "tier_label": label,
                "periods_today": teacher_day_count.get(k, 0), "periods_week": teacher_week_count.get(k, 0),
            }
            for k in sorted(tier_keys, key=lambda k: (teacher_day_count.get(k, 0), teacher_week_count.get(k, 0), display_name[k]))
        ]

    periods_out = []
    for ap in absent_periods:
        period, grade, section, subject = ap["period_number"], ap["grade_name"], ap["section_name"], ap["subject"]
        subj_norm = _norm(subject)

        free = [k for k in all_keys if period not in occupied_period.get(k, set())]

        tier1, tier2, tier3, tier4 = [], [], [], []
        for k in free:
            subs = teacher_subjects.get(k, set())
            grade_sections = teacher_grade_sections.get(k, set())
            grades_taught = teacher_grades.get(k, set())
            is_dmt_sports = any(any(kw in s for kw in DMT_SPORTS_KEYWORDS) for s in subs)
            if is_dmt_sports and (grade, section) in grade_sections:
                tier1.append(k)
            elif subj_norm in subs and (grade, section) in grade_sections:
                tier2.append(k)
            elif grade in grades_taught:
                tier3.append(k)
            else:
                tier4.append(k)

        suggestions = (
            build(tier1, 1, "DMT/Sports teacher for this class") +
            build(tier2, 2, "Same subject teacher for this class") +
            build(tier3, 3, "Teaches this grade elsewhere") +
            build(tier4, 4, "Other free teacher")
        )

        periods_out.append({
            "period_number": period, "grade_name": grade, "section_name": section,
            "subject": subject, "suggestions": suggestions,
        })

    return periods_out

File: Timetable/frontend-v2/src/lib/substitution.js
// Ranks substitute-teacher candidates for every period a specific teacher has
// on a given day, per the school's actual preference order (confirmed by the
// user 2026-07-31, deduped across tiers 2026-08-01):
//   1. Same grade, same subject (any other section)
//   2. Same grade, any other subject
//   3. "More" - next grade down, one grade per click, capped at 3 grades
//      below the absent period's grade
//   4. "More" - any free SPA teacher
//   5. "More" - any free PE teacher
//   6. "More" - any other free teacher
// Within a tier, the least-loaded candidate (fewest periods that day, then
// fewest that week) is listed first. A teacher only ever appears in the
// FIRST (highest-priority) tier they qualify for - once claimed there,
// later tiers exclude them, so the UI never has to de-dup on display and a
// name never appears twice across a period's suggestions.

const SPA_RE = /\bspa\b/i;
const PE_RE = /\bpe\b/i;

function norm(s) {
  return (s || "").trim().replace(/\s+/g, " ").toLowerCase();
}

function gradeNumber(gradeName) {
  const m = (gradeName || "").match(/\d+/);
  return m ? parseInt(m[0], 10) : null;
}

// lessons: the whole week's placed, teacher-assigned lessons - each
// {day_of_week, period_number, grade_name, section_name, subject, teacher_name}.
export function computeSuggestions(lessons, absentTeacherName, dayOfWeek, allTeacherNames) {
  const absentKey = norm(absentTeacherName);

  const displayName = new Map();
  for (const n of allTeacherNames) {
    if (n) displayName.set(norm(n), n);
  }
  for (const l of lessons) {
    if (l.teacher_name && !displayName.has(norm(l.teacher_name))) {
      displayName.set(norm(l.teacher_name), l.teacher_name);
    }
  }

  const occupiedPeriod = new Map(); // teacherKey -> Set(period_number) busy on dayOfWeek
  const teacherSubjects = new Map(); // teacherKey -> Set(subject, lowercased)
  const teacherGrades = new Map(); // teacherKey -> Set(grade_name)
  const teacherDayCount = new Map();
  const teacherWeekCount = new Map();

  for (const l of lessons) {
    if (!l.teacher_name) continue;
    const key = norm(l.teacher_name);
    if (!teacherSubjects.has(key)) teacherSubjects.set(key, new Set());
    teacherSubjects.get(key).add(norm(l.subject));
    if (!teacherGrades.has(key)) teacherGrades.set(key, new Set());
    teacherGrades.get(key).add(l.grade_name);
    teacherWeekCount.set(key, (teacherWeekCount.get(key) || 0) + 1);
    if (l.day_of_week === dayOfWeek) {
      if (!occupiedPeriod.has(key)) occupiedPeriod.set(key, new Set());
      occupiedPeriod.get(key).add(l.period_number);
      teacherDayCount.set(key, (teacherDayCount.get(key) || 0) + 1);
    }
  }

  const absentPeriods = lessons
    .filter((l) => norm(l.teacher_name) === absentKey && l.day_of_week === dayOfWeek)
    .slice()
    .sort((a, b) => a.period_number - b.period_number);

  const allKeys = [...displayName.keys()].filter((k) => k !== absentKey);

  function rank(keys) {
    return keys
      .slice()
      .sort((a, b) => {
        const da = teacherDayCount.get(a) || 0;
        const db_ = teacherDayCount.get(b) || 0;
        if (da !== db_) return da - db_;
        const wa = teacherWeekCount.get(a) || 0;
        const wb = teacherWeekCount.get(b) || 0;
        if (wa !== wb) return wa - wb;
        return displayName.get(a) < displayName.get(b) ? -1 : displayName.get(a) > displayName.get(b) ? 1 : 0;
      })
      .map((k) => ({
        teacher_name: displayName.get(k),
        periods_today: teacherDayCount.get(k) || 0,
        periods_week: teacherWeekCount.get(k) || 0,
      }));
  }

  // Distinct grade names actually present, so "3 grades below" can be
  // resolved to a real grade name (e.g. Grade 9 -> Grade 6), not just a
  // number that might not correspond to an actual grade in this data.
  const gradeNumToName = new Map();
  for (const l of lessons) {
    const gn = gradeNumber(l.grade_name);
    if (gn !== null && !gradeNumToName.has(gn)) gradeNumToName.set(gn, l.grade_name);
  }

  const periodsOut = [];
  for (const ap of absentPeriods) {
    const { period_number: period, grade_name: grade, section_name: section, subject } = ap;
    const subjNorm = norm(subject);
    const gradeNum = gradeNumber(grade);

    const free = allKeys.filter((k) => !(occupiedPeriod.get(k) || new Set()).has(period));

    // Each tier claims candidates from `free`, minus whoever an earlier
    // (higher-priority) tier already claimed - so every name appears at
    // most once across the whole period's suggestions.
    const claimed = new Set();
    function claim(keys) {
      const fresh = keys.filter((k) => !claimed.has(k));
      fresh.forEach((k) => claimed.add(k));
      return rank(fresh);
    }

    const sameGradeSameSubject = claim(
      free.filter((k) => (teacherGrades.get(k) || new Set()).has(grade) && (teacherSubjects.get(k) || new Set()).has(subjNorm))
    );
    const sameGradeOtherSubject = claim(free.filter((k) => (teacherGrades.get(k) || new Set()).has(grade)));

    const lowerGradeTiers = [];
    if (gradeNum !== null) {
      for (let down = 1; down <= 3; down++) {
        const lowerGradeName = gradeNumToName.get(gradeNum - down);
        if (!lowerGradeName) continue;
        const candidates = claim(free.filter((k) => (teacherGrades.get(k) || new Set()).has(lowerGradeName)));
        lowerGradeTiers.push({ gradeName: lowerGradeName, candidates });
      }
    }

    const spaTeachers = claim(free.filter((k) => [...(teacherSubjects.get(k) || new Set())].some((s) => SPA_RE.test(s))));
    const peTeachers = claim(free.filter((k) => [...(teacherSubjects.get(k) || new Set())].some((s) => PE_RE.test(s))));
    const otherTeachers = claim(free);

    periodsOut.push({
      period_number: period,
      grade_name: grade,
      section_name: section,
      subject,
      sameGradeSameSubject,
      sameGradeOtherSubject,
      lowerGradeTiers,
      spaTeachers,
      peTeachers,
      otherTeachers,
    });
  }

  return periodsOut;
}
